processor/src/test/java/org/mapstruct/ap/test/value/spi/CustomEnumNamingStrategyTest.java

```diff
@@ -0,0 +1,123 @@
+ */
+package org.mapstruct.ap.test.value.spi;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.WithServiceImplementation;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+import org.mapstruct.ap.testutil.runner.GeneratedSource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@RunWith(AnnotationProcessorTestRunner.class)
+@WithClasses({
+    CheeseType.class,
+    CustomCheeseType.class,
+    CustomEnumMarker.class,
+})
+@WithServiceImplementation(CustomEnumNamingStrategy.class)
+public class CustomEnumNamingStrategyTest {
+
+    @Rule
+    public final GeneratedSource generatedSource = new GeneratedSource();
+
+    @Test
+    @WithClasses({
+        CustomCheeseMapper.class
+    })
+    public void shouldApplyCustomEnumNamingStrategy() {
+        generatedSource.addComparisonToFixtureFor( CustomCheeseMapper.class );
+        CustomCheeseMapper mapper = CustomCheeseMapper.INSTANCE;
+
+        // CheeseType -> CustomCheeseType
+        assertThat( mapper.map( (CheeseType) null ) ).isEqualTo( CustomCheeseType.UNSPECIFIED );
+        assertThat( mapper.map( CheeseType.BRIE ) ).isEqualTo( CustomCheeseType.CUSTOM_BRIE );
+        assertThat( mapper.map( CheeseType.ROQUEFORT ) ).isEqualTo( CustomCheeseType.CUSTOM_ROQUEFORT );
+
+        // CustomCheeseType -> CheeseType
+        assertThat( mapper.map( (CustomCheeseType) null ) ).isNull();
+        assertThat( mapper.map( CustomCheeseType.UNSPECIFIED ) ).isNull();
+        assertThat( mapper.map( CustomCheeseType.CUSTOM_BRIE ) ).isEqualTo( CheeseType.BRIE );
+        assertThat( mapper.map( CustomCheeseType.CUSTOM_ROQUEFORT ) ).isEqualTo( CheeseType.ROQUEFORT );
+        assertThat( mapper.map( CustomCheeseType.UNRECOGNIZED ) ).isNull();
+
+        // CheeseType -> String
+        assertThat( mapper.mapToString( (CheeseType) null ) ).isNull();
+        assertThat( mapper.mapToString( CheeseType.BRIE ) ).isEqualTo( "BRIE" );
+        assertThat( mapper.mapToString( CheeseType.ROQUEFORT ) ).isEqualTo( "ROQUEFORT" );
+
+        // CustomCheeseType -> String
+        assertThat( mapper.mapToString( (CustomCheeseType) null ) ).isNull();
+        assertThat( mapper.mapToString( CustomCheeseType.UNSPECIFIED ) ).isNull();
+        assertThat( mapper.mapToString( CustomCheeseType.CUSTOM_BRIE ) ).isEqualTo( "BRIE" );
+        assertThat( mapper.mapToString( CustomCheeseType.CUSTOM_ROQUEFORT ) ).isEqualTo( "ROQUEFORT" );
+        assertThat( mapper.mapToString( CustomCheeseType.UNRECOGNIZED ) ).isNull();
+
+        // String - > CheeseType
+        assertThat( mapper.mapStringToCheese( null ) ).isNull();
+        assertThat( mapper.mapStringToCheese( "BRIE" ) ).isEqualTo( CheeseType.BRIE );
+        assertThat( mapper.mapStringToCheese( "ROQUEFORT" ) ).isEqualTo( CheeseType.ROQUEFORT );
```

Comment:
what would happen if in the "non custom" you would have a prefix `CUSTOM_` would a user expect this if it's the enum related to protobuf for instance?

Reply:
This is mapping `String` to `CheeseType`. `CheeseType` is a regular enum. The special enum is `CustomCheeseType` for that one you would expect the `String` `BRIE` to match the enum `CustomCheeseType.CUSTOM_BRIE`. Since the name based mapping constant for the `CustomCheeseType` is stripping the `CUSTOM_` prefix

Reply:
ah.. I get it.. Ok. nice one.. The user is in control.